### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [1]:
# Dependencies and Setup
import pandas as pd

# File to Load (Remember to Change These)
school_data_to_load = "Resources/schools_complete.csv"
student_data_to_load = "Resources/students_complete.csv"

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load)
student_data = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
merged_school_data_df = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])

## District Summary

* Calculate the total number of schools

* Calculate the total number of students

* Calculate the total budget

* Calculate the average math score 

* Calculate the average reading score

* Calculate the percentage of students with a passing math score (70 or greater)

* Calculate the percentage of students with a passing reading score (70 or greater)

* Calculate the percentage of students who passed math **and** reading (% Overall Passing)

* Create a dataframe to hold the above results

* Optional: give the displayed data cleaner formatting

In [2]:
#Calculate the total number of schools
total_schools  = merged_school_data_df["School ID"].value_counts().count()

#Calculate the total number of students
total_students = merged_school_data_df["Student ID"].count()

#Calculate the total budget
total_budget = school_data["budget"].sum()

#Calculate the average math score
avg_math_score = student_data["math_score"].mean()

#Calculate the average reading score
avg_read_score = student_data["reading_score"].mean()

#Calculate the percentage of students with a passing math score (70 or greater)
pct_std_pass_math = (student_data.loc[student_data["math_score"] >= 70].count() / total_students)[1]

#Calculate the percentage of students with a passing reading score (70 or greater)
pct_std_pass_reading = (student_data.loc[student_data["reading_score"] >= 70].count() / total_students)[1]

#Calculate the percentage of students who passed math and reading (% Overall Passing)
overall_pass = (student_data.loc[(student_data["math_score"] >= 70) & 
                                (student_data["reading_score"] >= 70)].count() / total_students)[1]

#Create a dataframe to hold the above results
df = pd.DataFrame([{"Total Schools"         : total_schools,
                    "Total Students"        : total_students,           
                    "Total Budget"          : total_budget,
                    "Average Math Score"    :  avg_math_score,
                    "Average Reading Score" :  avg_read_score,
                    "% Passing Math"        :  pct_std_pass_math    * 100,
                    "% Passing Reading"     :  pct_std_pass_reading * 100,
                    "% Overall Passing"     :  overall_pass         * 100}])

#Apply formatting
df["Total Schools"]         = df["Total Schools"].map("{:,}".format)
df["Total Students"]        = df["Total Students"].map("{:,}".format)
df["Total Budget"]           = df["Total Budget"].map("${:,.2f}".format)
df["Average Math Score"]    = df["Average Math Score"].map("{:.2f}".format)
df["Average Reading Score"] = df["Average Reading Score"].map("{:.2f}".format)
df["% Passing Math"]        = df["% Passing Math"].map("{:.2f}".format)
df["% Passing Reading"]     = df["% Passing Reading"].map("{:.2f}".format)
df["% Overall Passing"]     = df["% Overall Passing"].map("{:.2f}".format)

#Display visualization
df

Total Schools Total Students    Total Budget Average Math Score  \
0            15         39,170  $24,649,428.00              78.99   

  Average Reading Score % Passing Math % Passing Reading % Overall Passing  
0                 81.88          74.98             85.81             65.17

## School Summary

* Create an overview table that summarizes key metrics about each school, including:
  * School Name
  * School Type
  * Total Students
  * Total School Budget
  * Per Student Budget
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * % Overall Passing (The percentage of students that passed math **and** reading.)
  
* Create a dataframe to hold the above results

In [3]:
#SchoolType (School name comes for free)
school_type_series = school_data.set_index(["school_name"])["type"] #can't do a group by here

#Total Students
student_count_series = merged_school_data_df.groupby(["school_name"]).count()["Student ID"]

#Total School Budget
total_school_budget_series = merged_school_data_df.groupby(["school_name"]).mean()["budget"]

#Per Student Budget
per_student_budget_series = (total_school_budget_series/student_count_series)

#Average Math Score
avg_math_score_series = merged_school_data_df.groupby(["school_name"]).mean()["math_score"]

#Average Reading Score
avg_reading_score_series = merged_school_data_df.groupby(["school_name"]).mean()["reading_score"]

#% Passing Math
math_passing_series = (merged_school_data_df.loc[merged_school_data_df["math_score"] >= 70]).groupby(["school_name"])["Student ID"].count() / student_count_series 

#% Passing Reading
reading_passing_series = (merged_school_data_df.loc[merged_school_data_df["reading_score"] >= 70]).groupby(["school_name"])["Student ID"].count() / student_count_series 

#% Overall Passing (The percentage of students that passed math and reading.)
overall_passing_series = (merged_school_data_df.loc[(merged_school_data_df["math_score"] >= 70) & (merged_school_data_df["reading_score"] >= 70)]).groupby(["school_name"])["Student ID"].count() / student_count_series 

series_dictionary = {"School Type"          : school_type_series,
                     "Student Counts"       : student_count_series,
                     "Total School Budget"  : total_school_budget_series,
                     "Per Student Budget"   : per_student_budget_series,
                     "Average Math Score"   : avg_math_score_series,
                     "Average Reading Score": avg_reading_score_series,
                     "% Passing Math"       : math_passing_series    * 100,
                     "% Passing Reading"    : reading_passing_series * 100,
                     "% Overall Passing"    : overall_passing_series * 100}
   
#Convert dict to DataFrame
df = pd.DataFrame(series_dictionary)

#Apply formatting
df["Student Counts"]         = df["Student Counts"].map("{:,}".format)
df["Average Math Score"]     = df["Average Math Score"].map("{:.2f}".format)
df["Average Reading Score"]  = df["Average Reading Score"].map("{:.2f}".format)
df["Total School Budget"]    = df["Total School Budget"].map("${:,.2f}".format)
df["Per Student Budget"]     = df["Per Student Budget"].map("${:,.2f}".format)
df["% Passing Math"]         = df["% Passing Math"].map("{:.2f}".format)
df["% Passing Reading"]      = df["% Passing Reading"].map("{:.2f}".format)
df["% Overall Passing"]      = df["% Overall Passing"].map("{:.2f}".format)

#Clear index name
df.index.name = None

#Display Visualization
df

School Type Student Counts Total School Budget  \
Bailey High School       District          4,976       $3,124,928.00   
Cabrera High School       Charter          1,858       $1,081,356.00   
Figueroa High School     District          2,949       $1,884,411.00   
Ford High School         District          2,739       $1,763,916.00   
Griffin High School       Charter          1,468         $917,500.00   
Hernandez High School    District          4,635       $3,022,020.00   
Holden High School        Charter            427         $248,087.00   
Huang High School        District          2,917       $1,910,635.00   
Johnson High School      District          4,761       $3,094,650.00   
Pena High School          Charter            962         $585,858.00   
Rodriguez High School    District          3,999       $2,547,363.00   
Shelton High School       Charter          1,761       $1,056,600.00   
Thomas High School        Charter          1,635       $1,043,130.00   
Wilson High School        Charter          2,283       $1,319,574.00   
Wright High School        Charter          1,800       $1,049,400.00   

                      Per Student Budget Average Math Score  \
Bailey High School               $628.00              77.05   
Cabrera High School              $582.00              83.06   
Figueroa High School             $639.00              76.71   
Ford High School                 $644.00              77.10   
Griffin High School              $625.00              83.35   
Hernandez High School            $652.00              77.29   
Holden High School               $581.00              83.80   
Huang High School                $655.00              76.63   
Johnson High School              $650.00              77.07   
Pena High School                 $609.00              83.84   
Rodriguez High School            $637.00              76.84   
Shelton High School              $600.00              83.36   
Thomas High School               $638.00              83.42   
Wilson High School               $578.00              83.27   
Wright High School               $583.00              83.68   

                      Average Reading Score % Passing Math % Passing Reading  \
Bailey High School                    81.03          66.68             81.93   
Cabrera High School                   83.98          94.13             97.04   
Figueroa High School                  81.16          65.99             80.74   
Ford High School                      80.75          68.31             79.30   
Griffin High School                   83.82          93.39             97.14   
Hernandez High School                 80.93          66.75             80.86   
Holden High School                    83.81          92.51             96.25   
Huang High School                     81.18          65.68             81.32   
Johnson High School                   80.97          66.06             81.22   
Pena High School                      84.04          94.59             95.95   
Rodriguez High School                 80.74          66.37             80.22   
Shelton High School                   83.73          93.87             95.85   
Thomas High School                    83.85          93.27             97.31   
Wilson High School                    83.99          93.87             96.54   
Wright High School                    83.95          93.33             96.61   

                      % Overall Passing  
Bailey High School                54.64  
Cabrera High School               91.33  
Figueroa High School              53.20  
Ford High School                  54.29  
Griffin High School               90.60  
Hernandez High School             53.53  
Holden High School                89.23  
Huang High School                 53.51  
Johnson High School               53.54  
Pena High School                  90.54  
Rodriguez High School             52.99  
Shelton High School               89.89  
Thomas High School               

## Top Performing Schools (By % Overall Passing)

* Sort and display the top five performing schools by % overall passing.

In [4]:
#Sort bottom performing schools by % overall passing.
sorted_df = df.sort_values(by=['% Overall Passing'], ascending=False)

#Clear index name
sorted_df.index.name = None

#Display Visualization
sorted_df.head(5)

School Type Student Counts Total School Budget  \
Cabrera High School     Charter          1,858       $1,081,356.00   
Thomas High School      Charter          1,635       $1,043,130.00   
Griffin High School     Charter          1,468         $917,500.00   
Wilson High School      Charter          2,283       $1,319,574.00   
Pena High School        Charter            962         $585,858.00   

                    Per Student Budget Average Math Score  \
Cabrera High School            $582.00              83.06   
Thomas High School             $638.00              83.42   
Griffin High School            $625.00              83.35   
Wilson High School             $578.00              83.27   
Pena High School               $609.00              83.84   

                    Average Reading Score % Passing Math % Passing Reading  \
Cabrera High School                 83.98          94.13             97.04   
Thomas High School                  83.85          93.27             97.31   
Griffin High School                 83.82          93.39             97.14   
Wilson High School                  83.99          93.87             96.54   
Pena High School                    84.04          94.59             95.95   

                    % Overall Passing  
Cabrera High School             91.33  
Thomas High School              90.95  
Griffin High School             90.60  
Wilson High School              90.58  
Pena High School                90.54

## Bottom Performing Schools (By % Overall Passing)

* Sort and display the five worst-performing schools by % overall passing.

In [5]:
#Sort bottom performing schools by % overall passing.
sorted_df = df.sort_values(by=['% Overall Passing'], ascending=True)

#Clear index name
sorted_df.index.name = None

#Display Visualization
sorted_df.head(5)

School Type Student Counts Total School Budget  \
Rodriguez High School    District          3,999       $2,547,363.00   
Figueroa High School     District          2,949       $1,884,411.00   
Huang High School        District          2,917       $1,910,635.00   
Hernandez High School    District          4,635       $3,022,020.00   
Johnson High School      District          4,761       $3,094,650.00   

                      Per Student Budget Average Math Score  \
Rodriguez High School            $637.00              76.84   
Figueroa High School             $639.00              76.71   
Huang High School                $655.00              76.63   
Hernandez High School            $652.00              77.29   
Johnson High School              $650.00              77.07   

                      Average Reading Score % Passing Math % Passing Reading  \
Rodriguez High School                 80.74          66.37             80.22   
Figueroa High School                  81.16          65.99             80.74   
Huang High School                     81.18          65.68             81.32   
Hernandez High School                 80.93          66.75             80.86   
Johnson High School                   80.97          66.06             81.22   

                      % Overall Passing  
Rodriguez High School             52.99  
Figueroa High School              53.20  
Huang High School                 53.51  
Hernandez High School             53.53  
Johnson High School               53.54

## Math Scores by Grade

* Create a table that lists the average Reading Score for students of each grade level (9th, 10th, 11th, 12th) at each school.

  * Create a pandas series for each grade. Hint: use a conditional statement.
  
  * Group each series by school
  
  * Combine the series into a dataframe
  
  * Optional: give the displayed data cleaner formatting

In [6]:
#Create a pandas series for each grade and group each series by school by math score
mean_math_score_grade_9_series  = (merged_school_data_df.loc[merged_school_data_df["grade"] == '9th']).groupby(["school_name"])["math_score"].mean() 
mean_math_score_grade_10_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '10th']).groupby(["school_name"])["math_score"].mean() 
mean_math_score_grade_11_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '11th']).groupby(["school_name"])["math_score"].mean() 
mean_math_score_grade_12_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '12th']).groupby(["school_name"])["math_score"].mean() 

#Combine the series into a dataframe
series_dictionary =  {"9th":  mean_math_score_grade_9_series,
                      "10th": mean_math_score_grade_10_series,
                      "11th": mean_math_score_grade_11_series,
                      "12th": mean_math_score_grade_12_series}

#Convert dict to DataFrame
df = pd.DataFrame(series_dictionary)

#Apply formatting
df["9th"]  = df["9th"].map("{:.2f}".format)
df["10th"] = df["10th"].map("{:.2f}".format)
df["11th"] = df["11th"].map("{:.2f}".format)
df["12th"] = df["12th"].map("{:.2f}".format)

#Clear index name
df.index.name = None

#Display visualization
df

9th   10th   11th   12th
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

## Reading Score by Grade 

* Perform the same operations as above for reading scores

In [7]:
#Create a pandas series for each grade and group each series by school by reading score
mean_reading_score_grade_9_series  = (merged_school_data_df.loc[merged_school_data_df["grade"]  == '9th']).groupby(["school_name"])["reading_score"].mean() 
mean_reading_score_grade_10_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '10th']).groupby(["school_name"])["reading_score"].mean() 
mean_reading_score_grade_11_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '11th']).groupby(["school_name"])["reading_score"].mean() 
mean_reading_score_grade_12_series = (merged_school_data_df.loc[merged_school_data_df["grade"] == '12th']).groupby(["school_name"])["reading_score"].mean() 

#Combine the above Series into a DataFrame
series_dictionary =  {"9th":  mean_math_score_grade_9_series,
                      "10th": mean_math_score_grade_10_series,
                      "11th": mean_math_score_grade_11_series,
                      "12th": mean_math_score_grade_12_series}

#Convert dict to DataFrame
df = pd.DataFrame(series_dictionary)

#Apply formatting
df["9th"]  = df["9th"].map("{:.2f}".format)
df["10th"] = df["10th"].map("{:.2f}".format)
df["11th"] = df["11th"].map("{:.2f}".format)
df["12th"] = df["12th"].map("{:.2f}".format)

#Clear index name
df.index.name = None

#Display visualization
df

9th   10th   11th   12th
Bailey High School     77.08  77.00  77.52  76.49
Cabrera High School    83.09  83.15  82.77  83.28
Figueroa High School   76.40  76.54  76.88  77.15
Ford High School       77.36  77.67  76.92  76.18
Griffin High School    82.04  84.23  83.84  83.36
Hernandez High School  77.44  77.34  77.14  77.19
Holden High School     83.79  83.43  85.00  82.86
Huang High School      77.03  75.91  76.45  77.23
Johnson High School    77.19  76.69  77.49  76.86
Pena High School       83.63  83.37  84.33  84.12
Rodriguez High School  76.86  76.61  76.40  77.69
Shelton High School    83.42  82.92  83.38  83.78
Thomas High School     83.59  83.09  83.50  83.50
Wilson High School     83.09  83.72  83.20  83.04
Wright High School     83.26  84.01  83.84  83.64

## Scores by School Spending

* Create a table that breaks down school performances based on average Spending Ranges (Per Student). Use 4 reasonable bins to group school spending. Include in the table each of the following:
  * Average Math Score
  * Average Reading Score
  * % Passing Math
  * % Passing Reading
  * Overall Passing Rate (Average of the above two)

In [8]:
per_school_budget = merged_school_data_df.groupby(["school_name"]).mean()["budget"]

In [9]:
per_school_counts = merged_school_data_df["school_name"].value_counts()

In [10]:
per_school_capita = per_school_budget / per_school_counts

In [11]:
#4 resonable bins to group school
bins = [0,585, 630, 645, 680]

#labels
group_labels = ["<$585", "$585-630", "$630-645", "$645-680"]

#Perform merge
merged_school_data_df['bins'] = pd.cut(merged_school_data_df['budget'] / merged_school_data_df['size'], bins, labels = group_labels, include_lowest=True)

#Group by bins
binned_groups = merged_school_data_df.groupby('bins')

#Average Math Score Series
avg_math_score_series       = binned_groups['math_score'].mean()

#Average Reading Score Series
avg_reading_score_series    = binned_groups['reading_score'].mean()

#% Passing Math Series
pct_passing_math_series     = merged_school_data_df[merged_school_data_df['math_score']     >= 70].groupby('bins')['Student ID'].count()/binned_groups['Student ID'].count()

#% Passing Reading Series
pct_passing_reading_series  = merged_school_data_df[merged_school_data_df['reading_score']  >= 70].groupby('bins')['Student ID'].count()/binned_groups['Student ID'].count()

#Overall Passing Rate Series (Average of the above two) 
pct_overall_passing_series  = merged_school_data_df[(merged_school_data_df['reading_score'] >= 70) & (merged_school_data_df['math_score'] >= 70)].groupby('bins')['Student ID'].count()/binned_groups['Student ID'].count()

#Populate dict using above Series
series_dictionary = {"Average Math Score"     : avg_math_score_series,
                     "Average Reading Score"  : avg_reading_score_series,
                     "% Passing Math"         : pct_passing_math_series    * 100,
                     "% Passing Reading"      : pct_passing_reading_series * 100,
                     "% Overall Passing"      : pct_overall_passing_series * 100}
 
#Apply formatting
series_dictionary["Average Math Score"]    = series_dictionary["Average Math Score"].map("{:.2f}".format)
series_dictionary["Average Reading Score"] = series_dictionary["Average Reading Score"].map("{:.2f}".format)
series_dictionary["% Passing Math"]        = series_dictionary["% Passing Math"].map("{:.2f}".format)
series_dictionary["% Passing Reading"]     = series_dictionary["% Passing Reading"].map("{:.2f}".format) 
series_dictionary["% Overall Passing"]     = series_dictionary["% Overall Passing"].map("{:.2f}".format)
 
#Convert dict into DataFrame
df = pd.DataFrame(series_dictionary)

#Set index name
df.index.name = "Spending Ranges (Per Student)"

#Display visualization
df                                                                                                                                                                                 

Average Math Score Average Reading Score  \
Spending Ranges (Per Student)                                            
<$585                                      83.36                 83.96   
$585-630                                   79.98                 82.31   
$630-645                                   77.82                 81.30   
$645-680                                   77.05                 81.01   

                              % Passing Math % Passing Reading  \
Spending Ranges (Per Student)                                    
<$585                                  93.70             96.69   
$585-630                               79.11             88.51   
$630-645                               70.62             82.60   
$645-680                               66.23             81.11   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$585                                     90.64  
$585-630                                  70.94  
$630-645                                  58.84  
$645-680                                  53.53

## Scores by School Size

* Perform the same operations as above, based on school size.

In [12]:
#4 resonable bins to group school size
bins = [0, 1000, 2000, 5000]

#labels
group_labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

#Perform merge
merged_school_data_df['bins'] = pd.cut(merged_school_data_df['size'], bins, labels = group_labels)

#Group by bins
binned_groups = merged_school_data_df.groupby('bins')

#Student count series
student_count_series = binned_groups['Student ID'].count()

#Average Math Score Series
avg_math_score_series      = binned_groups['math_score'].mean()

#Average Reading Score Series
avg_reading_score_series   = binned_groups['reading_score'].mean()

#% Passing Math Series
pct_passing_math_series    = merged_school_data_df[merged_school_data_df['math_score']     >= 70].groupby('bins')['Student ID'].count()/student_count_series

#% Passing Reading Series
pct_passing_reading_series = merged_school_data_df[merged_school_data_df['reading_score']  >= 70].groupby('bins')['Student ID'].count()/student_count_series

#% Overall Passing Series
pct_overall_passing_series = merged_school_data_df[(merged_school_data_df['reading_score'] >= 70) & (merged_school_data_df['math_score'] >= 70)].groupby('bins')['Student ID'].count()/student_count_series

#Initialize dict using above Series
series_dictionary = {"Average Math Score"     : avg_math_score_series,
                     "Average Reading Score"  : avg_reading_score_series,
                     "% Passing Math"         : pct_passing_math_series    * 100,
                     "% Passing Reading"      : pct_passing_reading_series * 100,
                     "% Overall Passing"      : pct_overall_passing_series * 100 }

#Apply formatting
series_dictionary["Average Math Score"]    = series_dictionary["Average Math Score"].map("{:.2f}".format)
series_dictionary["Average Reading Score"] = series_dictionary["Average Reading Score"].map("{:.2f}".format)
series_dictionary["% Passing Math"]        = series_dictionary["% Passing Math"].map("{:.2f}".format)
series_dictionary["% Passing Reading"]     = series_dictionary["% Passing Reading"].map("{:.2f}".format) 
series_dictionary["% Overall Passing"]     = series_dictionary["% Overall Passing"].map("{:.2f}".format)

#Convert dict into DataFrame
df = pd.DataFrame(series_dictionary)

#Set index name
df.index.name = "School Size"

#Display Visualization
df                                       

Average Math Score Average Reading Score % Passing Math  \
School Size                                                                  
Small (<1000)                   83.83                 83.97          93.95   
Medium (1000-2000)              83.37                 83.87          93.62   
Large (2000-5000)               77.48                 81.20          68.65   

                   % Passing Reading % Overall Passing  
School Size                                             
Small (<1000)                  96.04             90.14  
Medium (1000-2000)             96.77             90.62  
Large (2000-5000)              82.13             56.57

## Scores by School Type

* Perform the same operations as above, based on school type

In [13]:
#Group data by school type
grouped_data = merged_school_data_df.groupby(["type"])

#Set student count as DataSeries
student_count_series_grp   = grouped_data['Student ID'].count()

#Average Math Score Series
avg_math_score_series      = grouped_data['math_score'].mean()

#Average Reading Score Series
avg_reading_score_series   = grouped_data['reading_score'].mean()

#% Passing Math Series
pct_passing_math_series    = merged_school_data_df[merged_school_data_df['math_score']     >= 70].groupby(["type"])['Student ID'].count()/student_count_series_grp

#% Passing Reading Series
pct_passing_reading_series = merged_school_data_df[merged_school_data_df['reading_score']  >= 70].groupby(["type"])['Student ID'].count()/student_count_series_grp

#% Overall Passing Series
pct_overall_passing_series = merged_school_data_df[(merged_school_data_df["reading_score"] >= 70) & (merged_school_data_df['math_score'] >= 70)].groupby(["type"])['Student ID'].count()/student_count_series_grp

#Initialize dict using above Series
series_dictionary = {"Average Math Score"     : avg_math_score_series,
                     "Average Reading Score"  : avg_reading_score_series,
                     "% Passing Math"         : pct_passing_math_series     * 100,
                     "% Passing Reading"      : pct_passing_reading_series  * 100,
                     "% Overall Passing"      : pct_overall_passing_series  * 100,}
    
#Apply formatting   
series_dictionary["Average Math Score"]    = series_dictionary["Average Math Score"].map("{:.2f}".format)
series_dictionary["Average Reading Score"] = series_dictionary["Average Reading Score"].map("{:.2f}".format)
series_dictionary["% Passing Math"]        = series_dictionary["% Passing Math"].map("{:.2f}".format)
series_dictionary["% Passing Reading"]     = series_dictionary["% Passing Reading"].map("{:.2f}".format) 
series_dictionary["% Overall Passing"]     = series_dictionary["% Overall Passing"].map("{:.2f}".format)
 
#Convert dict into DataFrame
df = pd.DataFrame(series_dictionary)

#Set index name
df.index.name = "School Type"

#Display Visualization
df 

Average Math Score Average Reading Score % Passing Math  \
School Type                                                           
Charter                  83.41                 83.90          93.70   
District                 76.99                 80.96          66.52   

            % Passing Reading % Overall Passing  
School Type                                      
Charter                 96.65             90.56  
District                80.91             53.70

In [14]:
##########################################
#                                        #
#   SUPPORTIVE WORK FOR CONSIDERATIONS   #
#         (WRITE-UP FOLLOWS)             #
#                                        #
##########################################

#Group data by school type
grouped_data = merged_school_data_df.groupby(["type"])

#School Scount By Type Series
school_data.set_index("type")
school_count_by_type_series = school_data.groupby(["type"])["school_name"].count()

#Set student count as DataSeries
student_count_series_grp   = grouped_data['Student ID'].count()

#Total Budget by School Type
avg_budget_score_series    = grouped_data['budget'].sum()  #Additional Column for further analysis

#Total Students by School Type
student_count_by_school_type = grouped_data['Student ID'].count()  #Additional Column for further analysis

#Average Math Score Series
avg_math_score_series      = grouped_data['math_score'].mean()

#Average Reading Score Series
avg_reading_score_series   = grouped_data['reading_score'].mean()

#% Passing Math Series
pct_passing_math_series    = merged_school_data_df[merged_school_data_df['math_score']     >= 70].groupby(["type"])['Student ID'].count()/student_count_series_grp

#% Passing Reading Series
pct_passing_reading_series = merged_school_data_df[merged_school_data_df['reading_score']  >= 70].groupby(["type"])['Student ID'].count()/student_count_series_grp

#% Overall Passing Series
pct_overall_passing_series = merged_school_data_df[(merged_school_data_df["reading_score"] >= 70) & (merged_school_data_df['math_score'] >= 70)].groupby(["type"])['Student ID'].count()/student_count_series_grp

#Initialize dict using above Series
series_dictionary = {"Total School Count"    : school_count_by_type_series,
                     "Total Student Count"   : student_count_by_school_type,
                     "Total Budget"          : avg_budget_score_series,
                     "Average Math Score"    : avg_math_score_series,
                     "Average Reading Score" : avg_reading_score_series,
                     "% Passing Math"        : pct_passing_math_series    * 100,
                     "% Passing Reading"     : pct_passing_reading_series * 100,
                     "% Overall Passing"     : pct_overall_passing_series * 100 }
    
#Apply formatting 
series_dictionary["Total School Count"]    = series_dictionary["Total School Count"].map("{:,}".format)
series_dictionary["Total Student Count"]   = series_dictionary["Total Student Count"].map("{:,}".format)
series_dictionary["Total Budget"]          = series_dictionary["Total Budget"].map("${:,.2f}".format)
series_dictionary["Average Math Score"]    = series_dictionary["Average Math Score"].map("{:.2f}".format)
series_dictionary["Average Reading Score"] = series_dictionary["Average Reading Score"].map("{:.2f}".format)
series_dictionary["% Passing Math"]        = series_dictionary["% Passing Math"].map("{:.2f}".format)
series_dictionary["% Passing Reading"]     = series_dictionary["% Passing Reading"].map("{:.2f}".format) 
series_dictionary["% Overall Passing"]     = series_dictionary["% Overall Passing"].map("{:.2f}".format)
 
#Convert dict into DataFrame
df = pd.DataFrame(series_dictionary)

#Set index name
df.index.name = "School Type"

#Display Visualization
df 

Total School Count Total Student Count        Total Budget  \
School Type                                                              
Charter                      8              12,194  $12,493,275,585.00   
District                     7              26,976  $70,439,053,973.00   

            Average Math Score Average Reading Score % Passing Math  \
School Type                                                           
Charter                  83.41                 83.90          93.70   
District                 76.99                 80.96          66.52   

            % Passing Reading % Overall Passing  
School Type                                      
Charter                 96.65             90.56  
District                80.91             53.70

Consideration #1: (See Above) Charter schools have better math and reading scores than district schools. Most notably, 
the overall passing rate for Charter Schools is far better at 90.56% than the rate of 53.70% for District Schools.  The above visualization expands on previous analysis by the addition of "School Count", "Student Count", and "Total Budget" columns to gain further insight into the base-line data. As presented above, the budget for Charter Schools is around 12 mil compared to a District School budget of around 75.0 mil yet test scores are lower. There appears to be an underlying issue that money alone cannot fix which leads to consideration #2...

Consideration #2: By analyzing the data in the columns "Total School Count" and "Total Student Count", one can observe that test scores go down as school size goes up. One may conclude that school size and test performance are inversely correlated. The number of students in the District school system stands at nearly 27,000 while the students in the charter system stands at a little over 12,000. Another interesting finding is that there is 1 less District School than Charter Schools. The issue potential overcrowding becomes apparent. In conclusion, because there are significantly more students in the District School system and 1 fewer school, test performance is diminished.  The addition of extra budgetary monies cannot alleviate the spacing / class size issue. The solution is to build more District Schools in order to increase the student-teacher ratio.